### PR TITLE
upgrade nginx OpenTelemetry and gRPC

### LINKED ISSUE #251: Nginx: half of trace tags null when using newer gRPC
**Describe your environment**
nginx:1.22.1-alpine docker container

**Steps to reproduce**
Managed to get the webserver/v1.0.3 tag to compile with gRPC v1.51.1 and opentelemetry-cpp v1.8.1.

```
FROM nginx:1.22.1-alpine as build-otel
RUN apk add --no-cache git curl make build-base cmake patch linux-headers curl-dev zlib-dev pcre-dev
RUN git clone --shallow-submodules --depth 1 --recurse-submodules -b v1.51.1 https://github.com/grpc/grpc \
  && cd grpc \
  && mkdir -p cmake/build \
  && cd cmake/build \
  && cmake \
    -DgRPC_INSTALL=ON \
    -DgRPC_BUILD_TESTS=OFF \
    -DCMAKE_INSTALL_PREFIX=/install \
    -DCMAKE_BUILD_TYPE=Release \
    -DgRPC_BUILD_GRPC_NODE_PLUGIN=OFF \
    -DgRPC_BUILD_GRPC_OBJECTIVE_C_PLUGIN=OFF \
    -DgRPC_BUILD_GRPC_PHP_PLUGIN=OFF \
    -DgRPC_BUILD_GRPC_PHP_PLUGIN=OFF \
    -DgRPC_BUILD_GRPC_PYTHON_PLUGIN=OFF \
    -DgRPC_BUILD_GRPC_RUBY_PLUGIN=OFF \
    ../.. \
  && make -j2 \
  && make install
RUN git clone --shallow-submodules --depth 1 --recurse-submodules -b v1.8.1 \
  https://github.com/open-telemetry/opentelemetry-cpp.git \
  && cd opentelemetry-cpp \
  && mkdir build \
  && cd build \
  && cmake -DCMAKE_BUILD_TYPE=Release \
    -DCMAKE_INSTALL_PREFIX=/install \
    -DCMAKE_PREFIX_PATH=/install \
    -DWITH_OTLP=ON \
    -DWITH_OTLP_GRPC=ON \
    -DWITH_OTLP_HTTP=OFF \
    -DWITH_ABSEIL=ON \
    -DBUILD_TESTING=OFF \
    -DWITH_EXAMPLES=OFF \
    -DCMAKE_POSITION_INDEPENDENT_CODE=ON \
    .. \
  && make -j2 \
  && make install
RUN git clone --depth 1 --branch webserver/v1.0.3 https://github.com/open-telemetry/opentelemetry-cpp-contrib.git \
  && cd opentelemetry-cpp-contrib/instrumentation/nginx \
  && mkdir -p build \
  && cd build \
  && cmake -DCMAKE_BUILD_TYPE=Release \
    -DCMAKE_PREFIX_PATH=/install \
    -DCMAKE_INSTALL_PREFIX=/usr/share/nginx/modules \
    .. \
  && make -j2 \
  && make install

FROM nginx:1.22.1-alpine
RUN apk add --no-cache libstdc++
COPY --from=build-otel /usr/share/nginx/modules/otel_ngx_module.so /usr/share/nginx/modules/otel_ngx_module.so
COPY nginx.conf /etc/nginx/nginx.conf
```

**What is the expected behavior?**
All trace tags have a value. Example:
| Tag | Value |
| -- | -- | 
| http.flavor | 1.1 |  
| http.host | localhost:8080 |  
| http.method | GET |  
| http.scheme | http |  
| http.server_name | localhost |  
| http.status_code | 200 |  
| http.target | / |  
| http.user_agent | curl/7.86.0 |  
| internal.span.format | proto |  
| net.host.port | 80 |  
| net.peer.ip | 172.19.0.1 |  
| net.peer.port | 55056 |  
| otel.library.name | nginx |  
| span.kind | server|
| telemetry.sdk.language | cpp |  
| telemetry.sdk.name | opentelemetry |  
| telemetry.sdk.version | 1.8.1 |

**What is the actual behavior?**
Half of the tags are null. Example:
| Tag | Value |
| -- | -- | 
| http.flavor |   |
| http.host | localhost:8080 |  
| http.method |   |  
| http.scheme |   |  
| http.server_name | localhost |  
| http.status_code |   |  
| http.target | / |  
| http.user_agent | curl/7.86.0 |  
| internal.span.format | proto |  
| net.host.port |   |  
| net.peer.ip |   |  
| net.peer.port |   |  
| otel.library.name | nginx |  
| span.kind | server |
| telemetry.sdk.language | cpp |  
| telemetry.sdk.name | opentelemetry |  
| telemetry.sdk.version | 1.8.1 |

**Additional context**
Works fine if compiled with gRPC v1.36.4 and no `-DWITH_ABSEIL=ON` passed to opentelemetry, as per your test flow. https://github.com/open-telemetry/opentelemetry-cpp-contrib/blob/57852b2d1629afb9ae0f6ca699a27d030163570d/instrumentation/nginx/test/instrumentation/lib/mix/tasks/dockerfiles.ex#L4

I see that this nginx module is technically abandonware now (per #229) so maybe the documentation needs a big warning put on it that only gRPC up to 1.36.4 is fully functional?

### PR DESCRIPTION
**Pull request by esigo**
Fixes #251
Upgrades gRPC to v1.49.2 and OpenTelemetry to v1.8.1.

### CODE REVIEW

**Review by marcalff — APPROVED** [review]
When moving an include file like:

```
#include <opentelemetry/exporters/otlp/otlp_grpc_exporter.h>
```

Please:

- Add a comment explaining why the header file needs to be in a given order, when it creates side effects
- Add an empty line after the include, so it is seen as a separate block, and won't be reordered by accident by anyone running clang-format

Looks ok otherwise.

**Review by lalitb — APPROVED** [review]

**Comment by esigo** [comment]
> When moving an include file like:
> 
> ```
> #include <opentelemetry/exporters/otlp/otlp_grpc_exporter.h>
> ```
> 
> Please:
> 
> * Add a comment explaining why the header file needs to be in a given order, when it creates side effects
> * Add an empty line after the include, so it is seen as a separate block, and won't be reordered by accident by anyone running clang-format
> 
> Looks ok otherwise.

Thanks, added the comment and new line.

File: instrumentation/nginx/src/otel_ngx_module.cpp [hunk 1]
```diff
@@ -1,3 +1,8 @@
+// otlp_grpc_exporter header has to be include before any other API header to
```
Comment (ThomsonTan) [inline instrumentation/nginx/src/otel_ngx_module.cpp:1]:
```suggestion
// otlp_grpc_exporter header has to be included before any other API header to 
```

File: instrumentation/nginx/src/otel_ngx_module.cpp [hunk 2]
```diff
@@ -1,3 +1,8 @@
+// otlp_grpc_exporter header has to be include before any other API header to 
+// avoid conflict between Abseil library and OpenTelemetry C++ absl::variant.
+// https://github.com/open-telemetry/opentelemetry-cpp/tree/main/examples/otlp#additional-notes-regarding-abseil-library
+#include <opentelemetry/exporters/otlp/otlp_grpc_exporter.h>
```
[the remark this answers is not in the document]
Comment (esigo) [reply instrumentation/nginx/src/otel_ngx_module.cpp:5]: Thanks, done

**Comment by eguzki** [comment]
This is actually fixing one issue that got me stuck, Looking forward to having a new tag to include this upgrade.

Thanks

**Pull request merged**